### DISCUSSION TITLE
Bidirectional and real time bluetooth communication between hubs

### DISCUSSION THREAD

**Discussion by massimiliano-mantione** [pybricks/support#890] [Ideas]
Pull request "[BLE broadcast: communication between pybricks hubs](https://github.com/pybricks/pybricks-micropython/pull/80)" implements a broadcast system.
It is also available for testing [here](https://pybricks.com/projects/tutorials/wireless/hub-to-hub/broadcast/).

It is an amazing work, but IMHO in its current form it has two issues (which are related):

- when connecting two Technic Hubs in a bidirectional way (two hubs, A and B, and two topics, A and B: hub A publishes into topic A and hub B publishes into topic B, both read each other's topic) they are not able to communicate properly (see [this](https://github.com/pybricks/pybricks-micropython/pull/80#issuecomment-1367108766) comment to the PR, the previous ones for demo code and the following ones for ideas)
- even when it works, the expected delays are in the tens of milliseconds range, which is likely not what a lot of users would expect

BTW, probably the issue reported [here](https://github.com/pybricks/support/discussions/280#discussioncomment-4531273) is the same one I am seeing.

I am moving the discussion here to see if I am the only user perceiving the problems, and to see if the project would like to have a more "real time" implementation of hub-to-hub communication, either as an evolution of the current PR or implemented with a different strategy.

The issue can be tricky because the current API is *very* ergonomic and simple to use.

The ideas I proposed would lead to a similar (almost identical) API for transmitting and receiving, but the initial setup would be more complex and more rigid.
Particularly, all hubs would need to share the list of all topics, or at least all publishers would need the full list of all publishers (so they can sync their radio writing windows).

I'd like to see if this (applying configuration complexity to gain communication determinism) is a tradeoff that the project would be willing to make.

**Comment by dlech** [comment]
> * even when it works, the expected delays are in the tens of milliseconds range, which is likely not what a lot of users would expect

I don't think anything faster than this is physically possible. In general advertisement intervals for BLE are 20ms or larger and connection intervals are 30ms or larger (unless the device is an HID device).



> I'd like to see if this (applying configuration complexity to gain communication determinism) is a tradeoff that the project would be willing to make.

Broadcast data is limited to 31 bytes, including overhead, so there really isn't any room for anything complex. The current protocol was chosen because it is compatible with the official LEGO firmware.

For something more deterministic, I think we would need make a connection between devices instead of using advertising data.

**Comment by massimiliano-mantione** [comment]
> For something more deterministic, I think we would need make a connection between devices instead of using advertising data.

Ok, I get it :-)

Where can I find some info on the actual cc2640 capabilities and the API we can use to control it so I can sketch a proposal for an API?
Is the data sheet the documentation I need, or does the project have other docs?

Thanks!!!

**Reply by dlech** [reply]
The firmware on the Bluetooth chip is customized by LEGO so is is not documented, but we have most of this API to work with: https://software-dl.ti.com/lprf/simplelink_cc26x2_latest/docs/ble5stack/vendor_specific_guide/BLE_Vendor_Specific_HCI_Guide/index.html

L2CAP seems like it would be nice for a lossless connection-based communication stream, but unfortunately it can't currently be used with Windows.

**Reply by massimiliano-mantione** [reply]
> The firmware on the Bluetooth chip is customized by LEGO so is is not documented, but we have most of this API to work with: ...

It seems to me that the API is defined in `lib/ble5stack/central`.
Is it correct? Are those the definitions I can use on a Technic Hub?

However, I cannot find anything referring to L2CAP operations there.

**Reply by dlech** [reply]
That is correct. I couldn't find  a complete open source library from TI, so some APIs may have to be added manually (and hope that the firmware on the Bluetooth chip actually supports it).

**Reply by massimiliano-mantione** [reply]
I have been reading the code, trying to understand how to make two Technic Hubs communicate using BLE characteristics.

The high-level API would be _logically_ similar to the [messaging](https://pybricks.com/ev3-micropython/messaging.html) API on the Mindstorm EV3, with bidirectional mailboxes.
However, in this case, the device with the central role should pair to a peripheral "on demand", as a result of python API calls, and not manually before the program starts like in the EV3.

Are there C code examples of using a Technic Hub with a central role, and pairing to other devices?
Has this ever been done at all?

Another question, this time about devices with peripheral role.
It seems to me that with the current firmware they wait for connections only _before_ the python program is started: once the program starts I cannot connect the web IDE to the brick anymore.
Did I understand correctly?

**Reply by dlech** [reply]
We currently don't require any authentication, so no pairing/bonding is required. Just scan and connect like we already do for the Powered Up remote control (hub is central, remote is peripheral).

> Did I understand correctly?

Yes, that is correct.

**Reply by massimiliano-mantione** [reply]
One more question about bluetooth connection handling.

I see that for now a hub cannot connect to more than one other "device hub".

However, I was trying to see if the following scenario is supported in principle:

- one PC connected to a technic hub (let's call it "HUB A"
- "HUB A" exchanging stdin-stdout data with the PC, and at the same timeconnected to "HUB B" where "HUB B" has device role (just like HUB A with the PC)
- "HUB B" exchanging stdin-stdout data with "HUB A" (HUB B is the device)

What it seems to me is that when a Bluetooth connection is established the `notification_handler` is set in `bluetooth_stm32_cc2640.c`.

As an example of a hub (HUB A) connecting to another LWP3 device (which in my scenario would be HUB B) I have looked at the code supporting the connection to a remote (`pb_type_pupdevices_remote.c`).
It seems to me that in this case the `notification_handler` is set to a different handler than the one used when the hub receives a connection from a PC.

So my question: is the code "ready" to support both connections at the same time?
I do not have a remote to test it :-)

**Reply by dlech** [reply]
Pybricks already has support for connecting to generic devices using the LWP3 protocol. But I have a feeling that when you said "LWP3", you really meant a Powered Up hub running Pybricks firmware rather than official LEGO firmware. This would just be a matter of implementing the Nordic UART service in the central role on the hub.

**Reply by massimiliano-mantione** [reply]
I am extra slow in working on this (this is really a _side_ project)!
You are right, what I want to do is to make the Technic Hub able to talk to a NUS peripheral (which, as you pointed out, means I have to implement the "NUS central" role for it).
The NUS peripheral could be just another Technic Hub.

Anyway, today I tried to make some progress, for real, and I got stuck when modifying the `pbdrv_bluetooth_scan_and_connect` function (I am only looking at it in the cc2040 implementation).

It seems to me that the function is LWP3-specific, and actually remote-specific: it has the assumption that the peripheral has the LWP3 service and characteristic (or, at least, this is the way it looks to me).

Before I do changes in the "wrong way": would it be ok to add one more argument to `pbdrv_bluetooth_scan_and_connect` , stating the handle of the service we are expected to connect to?

The rest would "flow from there".
Particularly, it should be possible to connect to both an LWP3 remote and a NUS peripheral at the same time, and a few FIXME about "match callback to handle" (especially about notifications) would be fixed.
Is this the right way to go?

**Reply by dlech** [reply]
> would it be ok to add one more argument to `pbdrv_bluetooth_scan_and_connect` , stating the handle of the service we are expected to connect to?

If we can be sure that nearly all devices (>80%) that implement NUS also advertise that service, then that would work. 

Otherwise we would need to make a more complex matching function to match advertising data from a variety of devices.

In either case, the function will need to be modified to properly decode arbitrary advertising data instead of assuming specific advertising data at specific locations in the packet.

**Reply by massimiliano-mantione** [reply]
> In either case, the function will need to be modified to properly decode arbitrary advertising data instead of assuming specific advertising data at specific locations in the packet.

Since "the simpler the better" and the function can already filter devices by name, would it be possible to:
- require that a device name is supplied if a NUS connection is asked for
- connect to a device with that name without looking for the service handle in the advertising packet
- handle that peripheral as a NUS anyway when writing and receiving notifications (if it does not support the NUS service it will not work, likely the peripheral will always reply with errors to write requests and certainly it will never send NUS data notifications but that's "ok")

This would be easier to implement as a proof of concept to see if the communication works and the latency is low enough for the application I have in mind :-)
I have no problem giving a known name to the technic hub I'll use as a peripheral.

Would it work?

**Reply by dlech** [reply]
That sounds like a good place to start.

**Comment by massimiliano-mantione** [comment]
Or maybe the cc2040 is limited, we want to keep things "simple", and we want to support only one connected peripheral at a time.
I would still need to tell `pbdrv_bluetooth_scan_and_connect` what service it should be looking for but I would not allow more than one connection.
Some static variable would still need to store which kind of device we are connected to but there would be a single callback and a single "remote_handle", but with a different name (maybe `peripheral_handle`?).

**Reply by dlech** [reply]
The simpler the better. :smile:

**Comment by dlech** [comment]
I've been doing some work on Bluetooth recently. For now just revisiting the ("not realtime") communication using advertising data. But while I was working on that, I learned a few things. One is that I don't think L2CAP channels are going to be an option as I hoped. I wrote some debug code to read `UTIL_BuildRevision()` on the city hub which gets info on how the firmware was configured on the Bluetooth chip. The `L2CAP_CO_CHANNELS` bit was not enabled so it makes me think we won't be able to open arbitrary L2CAP channels.

So, I think my plan for now is to move stdio away from the existing Nordic UART service so that it is freed up for hub-to-hub communication that is faster and has higher bandwidth that the advertising data. So there are two major tasks. 1. allowing a hub to have multiple connections as a peripheral and 2. modify the existing central code to be more generic to allow connecting to other services besides LWP3.
